File: sdk/Linux/cpp/src/api/navigation.cpp
// Copyright (c) 2026 WeGo Robotics. All rights reserved.

#include "api/detail.hpp"
#include "shalom/api.hpp"

namespace shalom {

std::string RobotApi::navigateTo(const Pose2D &goal, std::string *err)
{
    if (!api::detail::finite(goal))
        return api::detail::invalid(err, "goal must contain finite numbers");
    return client_.sendRequest("cmd/goto", "{\"x\":" + api::detail::number(goal.x)
                                         + ",\"y\":" + api::detail::number(goal.y)
                                         + ",\"theta\":" + api::detail::number(goal.theta) + "}", err);
}

std::string RobotApi::cancelNavigation(std::string *err)
{
    return client_.sendRequest("cmd/nav_cancel", "{}", err);
}

bool RobotApi::publishVelocity(const Twist2D &velocity, std::string *err)
{
    if (!api::detail::finite(velocity.vx) || !api::detail::finite(velocity.vy)
        || !api::detail::finite(velocity.wz)) {
        if (err)
            *err = "velocity must contain finite numbers";
        return false;
    }
    return client_.publish("cmd/cmd_vel", "{\"vx\":" + api::detail::number(velocity.vx)
                                      + ",\"vy\":" + api::detail::number(velocity.vy)
                                      + ",\"wz\":" + api::detail::number(velocity.wz) + "}", err);
}

}  // namespace shalom

File: sdk/Windows/python/shalom_sdk/api/navigation.py
"""Goal navigation and manual base-velocity commands."""

from __future__ import annotations

from ..types import Response
from ._base import _ApiBase


class NavigationApi(_ApiBase):
    def navigate_to(self, x: float, y: float, theta: float = 0.0,
                    timeout_s: float = 3.0) -> Response:
        self._finite(x, y, theta)
        return self._client.request("cmd/goto", {"x": x, "y": y, "theta": theta}, timeout_s)

    def cancel_navigation(self, timeout_s: float = 3.0) -> Response:
        return self._client.request("cmd/nav_cancel", timeout_s=timeout_s)

    def publish_velocity(self, vx: float, vy: float, wz: float) -> None:
        """Publish one manual velocity sample; callers must send it at 20 Hz."""
        self._finite(vx, vy, wz)
        self._client.publish("cmd/cmd_vel", {"vx": vx, "vy": vy, "wz": wz})
